File: prototypePython/guard.py
import pygame
from constants import PLAYER_SIZE,TILE_SIZE

def is_between(a, x, b):
    return min(a, b) <= x <= max(a, b)


class Guard(pygame.sprite.Sprite):
    #route = [ { "action":"go" , "time":1000 , "position" : [2,4]} , {"action":"wait", "time":2000}              ]
    def __init__(self,x,y,route:list[dict], *groups: pygame.sprite.AbstractGroup) -> None:
        super().__init__(*groups)
        self.surf = pygame.Surface((PLAYER_SIZE,PLAYER_SIZE))
        self.rect = self.surf.get_rect(topleft=(x*TILE_SIZE,y*TILE_SIZE))
        self.surf.fill((20,200,20))


        self.detector_surf = pygame.Surface((4*TILE_SIZE,2*TILE_SIZE),pygame.SRCALPHA)
        self.detector_surf.fill(pygame.Color(140,180,100,120))
        self.detector_rect = self.detector_surf.get_rect(midleft=self.rect.midright)
        self.route = route
        self.current_route_index = 0
        self.current_action = None
        self.current_route_data = None
        self.speed = 2
        self.truePos = [self.rect.x,self.rect.y]
        self.set_action(self.route[0])
    def set_action(self,route_action:dict):
        action = route_action['action']
        if action == "wait":
            self.current_route_data = {"cumulTime":0,"time":route_action['time']}
        elif action == "go":
            pos = [route_action["position"][0]*TILE_SIZE,route_action["position"][1]*TILE_SIZE]
            self.current_route_data = {"cumulTime":0,"time":route_action["time"],"position":pos,"initialPos":[self.rect.x,self.rect.y]}
            self.truePos = [self.rect.x,self.rect.y]
        else:
            print("error !")
            exit(1)
        self.current_action = action

    def next_action(self):
        self.current_route_index = (self.current_route_index + 1)% len(self.route)
        self.set_action(self.route[self.current_route_index])
        
    def update(self,dt:float):
        if self.current_action == "wait":
            self.current_route_data["cumulTime"] += dt
            if(self.current_route_data["cumulTime"]>self.current_route_data["time"]):
                self.next_action()
        elif self.current_action == "go":
            time_proportion = self.current_route_data["cumulTime"] / self.current_route_data["time"] 
            position_differenceX = self.current_route_data["position"][0] - self.current_route_data["initialPos"][0]
            position_differenceY = self.current_route_data["position"][1] - self.current_route_data["initialPos"][1]

            target = self.current_route_data["initialPos"][0] + (position_differenceX * time_proportion)

            self.truePos[0] = target if (is_between(self.rect.x,target,self.current_route_data["position"][0])) else self.current_route_data["position"][0]

            target = self.current_route_data["initialPos"][1] + (position_differenceY * time_proportion)

            self.truePos[1] = target if (is_between(self.rect.y,target,self.current_route_data["position"][1])) else self.current_route_data["position"][1]

 
            # if(int(self.truePos[0]) in range(self.rect.x,self.current_route_data["position"][0])):
            #     self.truePos[0] = self.current_route_data["position"][0]

            # if(int(self.truePos[1]) in range(self.rect.y,self.current_route_data["position"][1])):
            #     self.truePos[1] = self.current_route_data["position"][1]
            if(self.rect.x - self.truePos[0] != 0):
                if(self.rect.x < self.truePos[0]):
                    self.detector_rect.midleft=self.rect.midright
                elif(self.rect.x > self.truePos[0]):
                    self.detector_rect.midright=self.rect.midleft

            if(self.rect.y - self.truePos[1] != 0):
                if(self.rect.y < self.truePos[1]):
                    self.detector_rect.midtop=self.rect.midbottom
                elif(self.rect.y > self.truePos[1]):
                    self.detector_rect.midbottom=self.rect.midtop


            self.rect.x = self.truePos[0]
            self.rect.y = self.truePos[1]

            self.current_route_data["cumulTime"] += dt

            
            if (self.current_route_data["position"] == [self.rect.x,self.rect.y]):
                self.next_action()
        else:
            print("unknown action",self.current_action)
            exit(1)

    def render(self,surface:pygame.Surface):
        surface.blit(self.detector_surf,self.detector_rect)
        surface.blit(self.surf,self.rect)

File: local/guard.cc
#include "guard.h"
#include "wall.h"
#include <iostream>
#include <cmath>
bool is_between(float number, float lower,float upper){
	return ((unsigned)(number-lower) <= (upper-lower));
}


RouteAction  newRoute(actionType type,float time,gf::Vector2i position){
	return RouteAction(type,time,0.0,position);
}

Guard::Guard(gf::Vector2i grid_pos):
	shape(GUARD_SIZE),
	detectorShape(DETECTOR_SIZE){
	
    speed = 200;
    position = grid_pos * WALL_SIZE;
    spawn_position = position;
    rect = gf::RectF().fromPositionSize(position,GUARD_SIZE);  
    color = gf::Color::Red;
   	shape.setColor(color);
    shape.setAnchor(gf::Anchor::TopLeft);
    shape.setPosition(position);
    detectorShape.setColor(gf::Color4f({0.7,0.7,0,0.5}));
    detectorShape.setAnchor(gf::Anchor::TopLeft);
   	detectorShape.setPosition(position);   
	dirGuard=3; 		

}

void Guard::reset(){
	position = spawn_position;

    rect = gf::RectF().fromPositionSize(position,GUARD_SIZE);  
    shape.setPosition(position);
    detectorShape.setColor(gf::Color4f({0.7,0.7,0,0.5}));
   	detectorShape.setPosition(position);
   	detectorRect = detectorRect.fromPositionSize(detectorShape.getPosition(),DETECTOR_SIZE);

	route_index = -1;
	nextAction();//initialize first action
}

void Guard::setRoute(std::vector<RouteAction> new_route){
	route = new_route;
	route_index = -1;
	
}

gf::RectF * Guard::getRect(){
	return &detectorRect;
}

gf::Vector2f Guard::getPosition()const{
    return position;
}

void Guard::nextAction(){
	if(route.size()==0){return;}
	route_index ++;
	
   	if(route_index >= (int)route.size()){
   		route_index = 0;
   	}


    RouteAction & currentAction = route[route_index];
   	currentAction.cumulated_time = 0;
//
    // std::cout << "Route index : "<< route_index <<"\n";
	// std::cout << "New action ! "<< (currentAction.type == actionType::GO ?  "GO" : "WAIT") <<"\n";
	// std::cout << "time : " << currentAction.time << "\n";
	// std::cout << "Cumul time : " << currentAction.cumulated_time << "\n";

	//
   	switch(currentAction.type){
   		case actionType::WAIT:
   			shape.setColor(gf::Color::Azure);
   			break;
   	   	case actionType::GO:
   	   	   	shape.setColor(gf::Color::Red);
   	   	   	last_position = position;

   			break;
   		default:
   			break;
   	}

}

int Guard::getdirectionGuard(){
	return dirGuard;
}

int Guard::getGridPosY(){
	float y_pos;
	y_pos = position.y + GUARD_SIZE.y - 0.1;
	return (int) y_pos / WALL_SIZE.y;
}

bool Guard::isPrint(){
	return print;
}

void Guard::setPrint(){
	print=true;
}

void Guard::update(gf::Time time){
	print=false;
	float dt = time.asSeconds();
	gf::Vector2f target;
	float time_proportion,deltaX,deltaY,upper,lower;
	;RouteAction & currentAction = route[route_index];

    switch(currentAction.type){
    	case actionType::WAIT:
    		break;
    	case actionType::GO:
    		time_proportion = currentAction.cumulated_time / currentAction.time;
    		deltaX = (currentAction.grid_position.x * WALL_SIZE.x) - last_position.x;
    		deltaY = (currentAction.grid_position.y * WALL_SIZE.y) - last_position.y;

			target = {
				(float)round(last_position.x + deltaX * time_proportion ),
				(float)round(last_position.y + deltaY * time_proportion),

			};
			//std::cout << "target : " << target[0] << ", " << target[1] << std::endl;
			lower = std::min(last_position.x,currentAction.grid_position.x * WALL_SIZE.x);	
			upper = std::max(last_position.x,currentAction.grid_position.x * WALL_SIZE.x);	

    		if ( is_between(target.x,lower,upper)){
    		 	position.x =target.x;
			}else{
    			position.x =  (currentAction.grid_position.x * WALL_SIZE.x);
    		}
    		
    		position.y = last_position.y + deltaY * time_proportion;

			lower = std::min(last_position.y,currentAction.grid_position.y * WALL_SIZE.y);	
			upper = std::max(last_position.y,currentAction.grid_position.y * WALL_SIZE.y);	

			//CLAMP VALUES
    		if ( is_between(target.y,lower,upper)){
    		 	position.y =target.y;
			}else{
    			position.y =  (currentAction.grid_position.y * WALL_SIZE.y);
    		}

    	   	rect = rect.fromPositionSize(position,GUARD_SIZE);
    		shape.setPosition(position);

			detectorShape.setPosition(rect.getTopLeft());

			deltaX = round(deltaX);
			deltaY = round(deltaY);
			if (abs(deltaX) > abs(deltaY)){
				if (deltaX > 0 ){
					//going right
					detectorShape.move({0,+(GUARD_SIZE.y * 0.5) - (DETECTOR_SIZE.y * 0.5)});
					dirGuard=1;
				}else if (deltaX <0){
					//going left
					detectorShape.move({-DETECTOR_SIZE.x +GUARD_SIZE.x,+(GUARD_SIZE.y * 0.5) - (DETECTOR_SIZE.y * 0.5)});
					dirGuard=2;
				}
			}else{
				if(deltaY > 0){
					//going down
					detectorShape.move({(GUARD_SIZE.x * 0.5) - (DETECTOR_SIZE.x * 0.5),0});
					dirGuard=3;
				}else if(deltaY <0){
					//going up
					detectorShape.move({(GUARD_SIZE.x * 0.5) - (DETECTOR_SIZE.x * 0.5),-DETECTOR_SIZE.y + GUARD_SIZE.y});
					dirGuard=4;
				}
			}		

    		detectorRect = detectorRect.fromPositionSize(detectorShape.getPosition(),DETECTOR_SIZE);


    		break;
    	default:
    		break;
    }



	currentAction.cumulated_time += dt;

	//unchanged

    
    if (currentAction.cumulated_time >= currentAction.time){
    	//std::cout << "dt : "<< dt << std::endl;
		nextAction();
    }
}

void Guard::render(gf::RenderTarget& target){
	//target.draw(detectorShape);
    //target.draw(detectorShape);

    //target.draw(shape);
};

